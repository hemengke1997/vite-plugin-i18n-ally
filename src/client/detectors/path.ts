import { type Detector } from './types'

const regex = /\/([^\/]*)/g

export class Path implements Detector {
  name = 'path' as const
  resolveLng(options: { lookup: number }) {
    const { lookup: lookupFromPathIndex } = options
    if (typeof window === 'undefined') return undefined

    const language = window.location.pathname.match(regex)
    if (!Array.isArray(language)) return undefined

    const index = typeof lookupFromPathIndex === 'number' ? lookupFromPathIndex : 0
    return language[index]?.replace('/', '')
  }
  persistLng(lng: string, options: { lookup: number | string; languages: string[] }) {
    const { lookup: lookupPathIndex, languages } = options
    const index = typeof lookupPathIndex === 'number' ? lookupPathIndex : 0
    const currentURL = new URL(window.location.href)
    const language = currentURL.pathname.match(regex)
    if (!Array.isArray(language)) return
    if (language[index] === `/${lng}`) return
    const delIndex = languages.some((l) => language[index] === `/${l}`) ? 1 : 0
    language.splice(index, delIndex, `/${lng}`)
    currentURL.pathname = language.join('')
    window.history.replaceState({}, '', currentURL)
  }
}
