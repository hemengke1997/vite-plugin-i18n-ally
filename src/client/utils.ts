export function ensureArray(arr?: string[] | string) {
  if (!arr) return undefined
  return Array.isArray(arr) ? arr : [arr]
}

export function omit<T extends Record<string, any>, K extends keyof T>(obj: T, keys: K[]): Omit<T, K> {
  const clone = { ...obj }
  keys.forEach((key) => {
    delete clone[key]
  })
  return clone
}
