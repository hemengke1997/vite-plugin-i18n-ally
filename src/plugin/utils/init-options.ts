import { type I18nDetectorOptions } from '..'
import { getI18nAllyConfigByKey, readI18nAllyConfig } from './vscode-settings'

const DEFAULT_OPTIONS: I18nDetectorOptions = {
  localesPaths: ['./src/locales', './locales'],
  root: process.cwd(),
  namespace: false,
  dotVscodePath: process.cwd(),
}

function getDefaultOptions(options?: I18nDetectorOptions): I18nDetectorOptions {
  let i18AllyConfig: Record<string, any> | undefined = undefined
  if (options?.dotVscodePath !== false) {
    // detect vscode settings of i18n-ally
    i18AllyConfig = readI18nAllyConfig(options?.dotVscodePath || (DEFAULT_OPTIONS.dotVscodePath as string))
  }

  if (i18AllyConfig) {
    return {
      localesPaths: getI18nAllyConfigByKey(i18AllyConfig, 'localesPaths') ?? DEFAULT_OPTIONS.localesPaths,
      pathMatcher: getI18nAllyConfigByKey(i18AllyConfig, 'pathMatcher'),
      namespace: getI18nAllyConfigByKey(i18AllyConfig, 'namespace') ?? DEFAULT_OPTIONS.namespace,
      ...DEFAULT_OPTIONS,
    }
  }

  return DEFAULT_OPTIONS
}

export function initOptions(options?: I18nDetectorOptions) {
  return {
    ...getDefaultOptions(options),
    ...options,
  } as Required<I18nDetectorOptions>
}
