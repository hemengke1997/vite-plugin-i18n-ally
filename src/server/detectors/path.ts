import type { Detector } from './types'

const regex = /\/([^/]*)/g

export class Path implements Detector {
  name = 'path' as const
  resolveLng(options: { lookup?: number, request: Request }) {
    const { lookup: lookupFromPathIndex, request } = options
    const language = new URL(request.url).pathname.match(regex)
    if (!Array.isArray(language))
      return undefined

    const index = typeof lookupFromPathIndex === 'number' ? lookupFromPathIndex : 0
    return language[index]?.replace('/', '')
  }
}
