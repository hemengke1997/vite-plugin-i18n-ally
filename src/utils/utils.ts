export function ensureArray(arr?: string[] | string) {
  if (!arr)
    return undefined
  return Array.isArray(arr) ? arr : [arr]
}

export function omit<T extends Record<string, any>, K extends keyof T>(obj: T, keys: K[]): Omit<T, K> {
  const clone = { ...obj }
  keys.forEach((key) => {
    delete clone[key]
  })
  return clone
}

/**
 * find。从 source 中找到 target 中的任意一个
 *
 * 如果开启lowerCase小写，则忽略大小写查找
 */
export function findByCase(source: string[], target: string | string[], lowerCase: boolean | undefined) {
  const targets = ensureArray(target) || []

  return source.find(item => targets.some(t => (lowerCase ? item.toLowerCase() === t.toLowerCase() : item === t)))
}

/**
 * 如果开启lowerCase小写，则将语言转为小写
 */
export function formatLng<T>(lng: T, lowerCaseLng: boolean | undefined): T {
  if (Array.isArray(lng)) {
    return lng.map((l: string | undefined) => (lowerCaseLng ? l?.toLowerCase() : l)) as T
  }
  if (lowerCaseLng) {
    return (lng as string | undefined)?.toLowerCase() as T
  }
  return lng
}
