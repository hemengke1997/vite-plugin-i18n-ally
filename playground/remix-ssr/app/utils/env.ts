import { Env } from 'vite-config-preset/client'

export function getEnv(): string {
  return import.meta.env.MODE
}

export function isDev(): boolean {
  return getEnv() === Env.development
}

export function isTest(): boolean {
  return getEnv() === Env.test
}

export function isProd(): boolean {
  return getEnv() === Env.production
}
