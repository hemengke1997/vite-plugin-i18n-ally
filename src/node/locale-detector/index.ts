import { cloneDeep, throttle, trimEnd, uniq } from 'es-toolkit'
import fg from 'fast-glob'
import tags from 'language-tags'
import fs from 'node:fs'
import path from 'node:path'
import { type ChangeEvent } from 'rollup'
import { normalizePath } from 'vite'
import { DefaultEnabledParsers } from '../parsers'
import { Parser } from '../parsers/Parser'
import { ParsePathMatcher } from '../path-matcher'
import { type I18nAllyOptions } from '../types'
import { debug } from '../utils/debugger'
import { unflatten } from '../utils/flat'
import { logger } from '../utils/logger'
import { VirtualModule } from '../utils/virtual'

const THROTTLE_DELAY = 1500

export type Config = Omit<Required<I18nAllyOptions>, 'useVscodeI18nAllyConfig'>

type PathMatcherType = RegExp

type DirStructure = 'file' | 'dir'

export interface FileInfo {
  filepath: string
  deepDirpath: string
  dirpath: string
  locale: string
  mtime: number
  namespace?: string
  matcher?: string
}

export interface ParsedFile extends FileInfo {
  value: Record<string, any>
}

export class LocaleDetector {
  readonly config: Config
  private _dirStructure: DirStructure = 'file'
  private _pathMatcher: { regex: PathMatcherType; matcher: string } | undefined
  private _localesPaths: string[]
  private _rootPath: string
  private _namespace: boolean

  public separator = '__'

  private _localeDirs: string[] = []
  private _files: Record<string, ParsedFile> = {}
  private _localeModules: {
    modules: Record<string, any>
    modulesWithNamespace: Record<string, any>
    virtualModules: Record<string, any>
    resolvedIds: Map<string, string>
  } = { modules: {}, modulesWithNamespace: {}, virtualModules: {}, resolvedIds: new Map() }

  constructor(c: Config) {
    this.config = c
    this._rootPath = c.root
    this._namespace = c.namespace
    const pathMatcher = c.pathMatcher
    if (pathMatcher) {
      this._pathMatcher = {
        regex: ParsePathMatcher(pathMatcher, this.enabledParserExts()),
        matcher: pathMatcher,
      }
    }

    this._localesPaths = c.localesPaths.map((item) =>
      trimEnd(normalizePath(path.isAbsolute(item) ? item : path.resolve(this._rootPath, item)), '/\\').replaceAll(
        '\\',
        '/',
      ),
    )
  }

  async init() {
    if (await this.findLocaleDirs()) {
      debug(`Initializing loader "${this._rootPath}"`)

      if (this._pathMatcher) {
        debug(`Custom Path Matcher: ${this._pathMatcher.matcher}`)
        debug(`Path Matcher Regex: ${this._pathMatcher.regex}`)
      } else {
        const pathMatcherGussed = await this.resolvePathMatcherByDirStructure()
        this._pathMatcher = {
          regex: ParsePathMatcher(pathMatcherGussed, this.enabledParserExts()),
          matcher: pathMatcherGussed,
        }
        debug(`Path Matcher: ${this._pathMatcher.matcher}`)
      }

      debug(`\nLocale Detector options: `, {
        root: this._rootPath,
        localesPaths: this._localesPaths,
        pathMatcher: this._pathMatcher.matcher,
        parserPlugins: this.getParsers(),
        namespace: this._namespace,
      })

      await this.loadAll()
    }

    this.update()
  }

  private update() {
    debug('Loading finished')
    this.updateLocaleModule()
  }

  private updateLocaleModule() {
    const modules: Record<string, any> = {}
    const modulesWithNamespace: Record<string, any> = {}
    this.files.forEach((file) => {
      const keyArray = [file.locale, file.namespace].filter(Boolean) as string[]
      keyArray.reduce((current, key, i) => {
        if (!current[key]) {
          current[key] = {}
        }
        if (i === keyArray.length - 1) {
          current[key] = {
            ...current[key],
            ...file.value,
          }
        }
        return current[key]
      }, modules)
    })

    if (this._namespace) {
      Object.keys(modules).forEach((locale) => {
        const value = modules[locale]
        Object.keys(value).forEach((namespace) => {
          const key = `${locale}${this.separator}${namespace}`
          modulesWithNamespace[key] = value[namespace]
        })
      })
    }

    const resolvedIds = new Map<string, string>()

    const virtualModules = cloneDeep({
      ...modules,
      ...modulesWithNamespace,
    })

    Object.keys(virtualModules).forEach((k) => {
      const id = VirtualModule.id(k)
      virtualModules[id] = virtualModules[k]
      resolvedIds.set(id, id)
      delete virtualModules[k]
    })

    this._localeModules = {
      modules,
      modulesWithNamespace,
      virtualModules,
      resolvedIds,
    }
  }

  get localeModules() {
    return this._localeModules
  }

  get allLocaleDirs() {
    return new Set(this.files.map((t) => t.deepDirpath))
  }

  get allLocaleFiles() {
    return new Set(this.files.map((t) => t.filepath))
  }

  async onFileChanged(event: ChangeEvent, { fsPath: filepath }: { fsPath: string }) {
    filepath = path.resolve(filepath)

    // not tracking
    if (event !== 'create' && !this._files[filepath]) {
      return
    }

    // already up-to-date
    if (event !== 'update' && this._files[filepath]?.mtime === this.getMtime(filepath)) {
      debug(`Skipped on loading "${filepath}" (same mtime)`)
      return
    }

    const { dirpath, relative } = this.getRelativePath(filepath) || {}

    if (!dirpath || !relative) {
      return
    }

    debug(`File changed (${event}) ${relative}`)

    switch (event) {
      case 'delete':
        this.unsetFile(filepath)
        this.throttledUpdate()
        break
      case 'create':
      case 'update':
        this.throttledLoadFile(dirpath, relative)
        break
    }

    return relative
  }

  private throttledUpdate = throttle(() => {
    this.update()
  }, THROTTLE_DELAY)

  private throttledLoadFileWaitingList: [string, string][] = []

  private throttledLoadFileExecutor = throttle(async () => {
    const list = this.throttledLoadFileWaitingList
    this.throttledLoadFileWaitingList = []
    if (list.length) {
      let changed = false
      for (const [d, r] of list) changed = (await this.loadFile(d, r)) || changed

      if (changed) this.update()
    }
  }, THROTTLE_DELAY)

  private throttledLoadFile = (d: string, r: string) => {
    if (!this.throttledLoadFileWaitingList.find(([a, b]) => a === d && b === r))
      this.throttledLoadFileWaitingList.push([d, r])
    this.throttledLoadFileExecutor()
  }

  private getRelativePath(filepath: string) {
    let dirpath = this._localeDirs.find((dir) => filepath.startsWith(dir))
    if (!dirpath) {
      return
    }

    let relative = path.relative(dirpath, filepath)

    if (process.platform === 'win32') {
      relative = relative.replaceAll('\\', '/')
      dirpath = dirpath.replaceAll('\\', '/')
    }

    return { dirpath, relative }
  }

  get files() {
    return Object.values(this._files)
  }

  private async loadAll() {
    for (const pathname of this._localeDirs) {
      try {
        debug(`Loading locales under ${pathname}`)
        await this.loadDirectory(pathname)
      } catch (e) {
        console.error(e)
      }
    }
    if (!this.files.length) {
      logger.warn(`No locale files detected.\n`, {
        timestamp: true,
      })
    }
  }

  private async loadDirectory(searchingPath: string) {
    const files = await fg('**/*.*', {
      cwd: searchingPath,
      onlyFiles: true,
      ignore: ['node_modules/**'],
      deep: undefined,
    })

    debug(`ðŸ“‚ Loading ${files.length} files under ${searchingPath}`)

    for (const relative of files) {
      await this.loadFile(searchingPath, relative)
    }
  }

  private async loadFile(dirpath: string, relativePath: string) {
    try {
      const result = this.getFileInfo(dirpath, relativePath)
      if (!result) {
        return
      }
      const { locale, parser, namespace, fullpath: filepath, matcher } = result
      if (!parser) {
        return
      }
      if (!locale) {
        return
      }

      const mtime = this.getMtime(filepath)

      debug(`ðŸ“‘ Loading (${locale}) ${relativePath} [${mtime}]`)

      const data = await parser.load(filepath)

      // support `i18n-ally.keyStyle` both `nesetd` and `flat`
      const value = unflatten(data)

      const deepDirpath = path.dirname(filepath)

      this._files[filepath] = {
        filepath,
        deepDirpath,
        dirpath,
        locale,
        mtime,
        value,
        namespace,
        matcher,
      }

      return true
    } catch (e) {
      this.unsetFile(relativePath)
      debug(`Failed to load ${e}`)
      console.error(e)
    }
  }

  private unsetFile(filepath: string) {
    delete this._files[filepath]
  }

  private getFileInfo(dirpath: string, relativePath: string) {
    const fullpath = path.resolve(dirpath, relativePath)

    debug(`Full path: ${fullpath}`)
    const ext = path.extname(relativePath)

    let match: RegExpExecArray | null = null
    let matcher: string | undefined

    match = this._pathMatcher!.regex.exec(relativePath)

    if (match && match.length > 0) {
      matcher = this._pathMatcher!.matcher
    }

    if (!match || match.length < 1) {
      return
    }

    debug(`Path Match:`, match)

    let namespace = match.groups?.namespace
    if (namespace) {
      namespace = namespace.replaceAll('/', '.')
    }

    const locale = match.groups?.locale

    if (!locale) {
      return
    }

    const parser = this.getMatchedParser(ext)

    const result = {
      locale,
      parser,
      ext,
      namespace,
      fullpath,
      matcher,
    }

    return result
  }

  private getMatchedParser(ext: string) {
    if (!ext.startsWith('.') && ext.includes('.')) {
      ext = path.extname(ext)
    }

    // resolve parser
    return this.getParsers().find((parser) => parser.supports(ext))
  }

  private enabledParserExts() {
    return this.getParsers()
      .map((item) => item.ext)
      .filter(Boolean)
      .join('|')
  }

  private getParsers() {
    const _parsers = this.config.parserPlugins?.filter(Boolean)

    const parsers = DefaultEnabledParsers
    if (_parsers?.length) {
      parsers.push(..._parsers.filter(Boolean).map((parser) => new Parser(parser!)))
    }

    return parsers
  }

  get pathMatcher() {
    return this._pathMatcher
  }

  get localeDirs() {
    return this._localeDirs
  }

  private getMtime(filepath: string) {
    try {
      return fs.statSync(filepath).mtimeMs
    } catch {
      return 0
    }
  }

  async findLocaleDirs() {
    this._files = Object.create(null)
    this._localeModules = Object.create(null)
    this._localeDirs = []

    if (this._localesPaths?.length) {
      try {
        const _locale_dirs = await fg(this._localesPaths, {
          cwd: this._rootPath,
          onlyDirectories: true,
        })

        if (this._localesPaths.includes('.')) {
          _locale_dirs.push('.')
        }

        this._localeDirs = uniq(_locale_dirs.map((p) => path.resolve(this._rootPath, p)))
        debug(`ðŸ“‚ Locale directories:`, this._localeDirs)
      } catch (e) {
        console.error(e)
      }
    }
    if (this._localeDirs.length === 0) {
      logger.warn(`No locales paths.\n`, {
        timestamp: true,
      })
      return false
    }

    return true
  }

  async guessDirStructure(): Promise<DirStructure> {
    const POSITIVE_RATE = 0.6

    const dir = this._localeDirs[0]

    const dirnames = await fg('*', {
      onlyDirectories: true,
      cwd: dir,
      deep: 1,
    })

    const total = dirnames.length
    if (total === 0) return 'file'

    const positives = dirnames.map((d) => tags.check(d))

    const positive = positives.filter((d) => d).length

    // if there are some dirs are named as locale code, guess it's dir mode
    return positive / total >= POSITIVE_RATE ? 'dir' : 'file'
  }

  async resolvePathMatcherByDirStructure() {
    this._dirStructure = await this.guessDirStructure()
    debug(`Directory structure: ${this._dirStructure}`)
    return this.resolvePathMatcher(this._dirStructure)
  }

  resolvePathMatcher(dirStructure?: DirStructure): string {
    if (dirStructure === 'file') {
      return '{locale}.{ext}'
    } else if (this._namespace) {
      return '{locale}/**/{namespace}.{ext}'
    } else {
      return '{locale}/**/*.{ext}'
    }
  }
}
