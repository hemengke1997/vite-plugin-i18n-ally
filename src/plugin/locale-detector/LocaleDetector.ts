import cloneDeep from 'clone-deep'
import fg from 'fast-glob'
import tags from 'language-tags'
import path from 'node:path'
import trimEnd from 'string.prototype.trimend'
import uniq from 'uniq'
import { normalizePath } from 'vite'
import { type I18nAllyOptions } from '../interface'
import { DefaultEnabledParsers } from '../parsers'
import { Parser } from '../parsers/Parser'
import { ParsePathMatcher } from '../path-matcher/PathMatcher'
import { PKGNAME, VIRTUAL } from '../utils/constant'
import { debug } from '../utils/debugger'
import { logger } from '../utils/logger'

export type Config = Omit<Required<I18nAllyOptions>, 'useVscodeI18nAllyConfig'>

type PathMatcherType = RegExp

type DirStructure = 'file' | 'dir'

export interface FileInfo {
  filepath: string
  deepDirpath: string
  dirpath: string
  locale: string
  namespace?: string
  matcher?: string
}

export interface ParsedFile extends FileInfo {
  value: object
}

export class LocaleDetector {
  readonly config: Config
  private _dirStructure: DirStructure = 'file'
  private _pathMatcher: { regex: PathMatcherType; matcher: string } | undefined
  private _localesPaths: string[]
  private _rootPath: string
  private _namespace: boolean

  private _localeDirs: string[] = []
  private _files: Record<string, ParsedFile> = {}
  private _localeModules: {
    modules: Record<string, any>
    virtualModules: Record<string, any>
    resolvedIds: Map<string, string>
  } = { modules: {}, virtualModules: {}, resolvedIds: new Map() }

  constructor(c: Config) {
    this.config = c
    this._rootPath = c.root
    this._namespace = c.namespace
    const pathMatcher = c.pathMatcher
    if (pathMatcher) {
      this._pathMatcher = {
        regex: ParsePathMatcher(pathMatcher, this.enabledParserExts()),
        matcher: pathMatcher,
      }
    }

    this._localesPaths = c.localesPaths.map((item) =>
      trimEnd(normalizePath(path.isAbsolute(item) ? item : path.resolve(this._rootPath, item)), '/\\').replaceAll(
        '\\',
        '/',
      ),
    )
  }

  async init() {
    if (await this.findLocaleDirs()) {
      debug(`Initializing loader "${this._rootPath}"`)

      if (this._pathMatcher) {
        debug(`Custom Path Matcher: ${this._pathMatcher.matcher}`)
        debug(`Path Matcher Regex: ${this._pathMatcher.regex}`)
      } else {
        const pathMatcherGussed = await this.resolvePathMatcherByDirStructure()
        this._pathMatcher = {
          regex: ParsePathMatcher(pathMatcherGussed, this.enabledParserExts()),
          matcher: pathMatcherGussed,
        }
        debug(`Path Matcher: ${this._pathMatcher.matcher}`)
      }

      debug(`\nThe real I18nAlly options: `, {
        root: this._rootPath,
        localesPaths: this._localesPaths,
        pathMatcher: this._pathMatcher.matcher,
        parserPlugins: this.getParsers(),
        namespace: this._namespace,
      })

      await this.loadAll()
    }

    this.update()
  }

  private update() {
    debug('Loading finished')
    this.updateLocaleModule()
  }

  private updateLocaleModule() {
    const modules: Record<string, any> = {}
    this.files.forEach((file) => {
      const keyArray = [file.locale, file.namespace].filter(Boolean) as string[]
      keyArray.reduce((current, key, i) => {
        if (!current[key]) {
          current[key] = {}
        }
        if (i === keyArray.length - 1) {
          current[key] = {
            ...current[key],
            ...file.value,
          }
        }
        return current[key]
      }, modules)
    })

    const resolvedIds = new Map<string, string>()

    const virtualModules = cloneDeep(modules)

    Object.keys(virtualModules).forEach((k) => {
      const id = `${VIRTUAL}-${k}`
      virtualModules[id] = virtualModules[k]
      resolvedIds.set(path.resolve(id), id)
      delete virtualModules[k]
    })

    this._localeModules = {
      modules,
      virtualModules,
      resolvedIds,
    }

    debug('Module locale updated', this._localeModules)
  }

  get localeModules() {
    return this._localeModules
  }

  get allLocaleDirs() {
    return new Set(this.files.map((t) => t.deepDirpath))
  }

  get allLocaleFiles() {
    return new Set(this.files.map((t) => t.filepath))
  }

  async onFileChanged({ fsPath: filepath }: { fsPath: string }) {
    filepath = path.resolve(filepath)

    const { dirpath, relative } = this.getRelativePath(filepath) || {}

    if (!dirpath || !relative) {
      return
    }

    debug(`File changed  ${relative} ${dirpath}`)

    return this.lazyLoadFile(dirpath, relative)
  }

  private loadFileWaitingList: [string, string][] = []

  get loadFileWaitingListLength() {
    return this.loadFileWaitingList.length
  }

  private loadFileExecutor = async () => {
    const list = this.loadFileWaitingList
    this.loadFileWaitingList = []
    if (list.length) {
      let changed = false
      for (const [d, r] of list) changed = (await this.loadFile(d, r)) || changed

      if (changed) {
        this.update()
        return true
      }
    }
    return false
  }

  private lazyLoadFile = async (d: string, r: string) => {
    if (!this.loadFileWaitingList.some(([a, b]) => a === d && b === r)) {
      this.loadFileWaitingList.push([d, r])
    }
    const updated = await this.loadFileExecutor()

    return updated
  }

  private getRelativePath(filepath: string) {
    let dirpath = this._localeDirs.find((dir) => filepath.startsWith(dir))
    if (!dirpath) {
      return
    }

    let relative = path.relative(dirpath, filepath)

    if (process.platform === 'win32') {
      relative = relative.replaceAll('\\', '/')
      dirpath = dirpath.replaceAll('\\', '/')
    }

    return { dirpath, relative }
  }

  get files() {
    return Object.values(this._files)
  }

  private async loadAll() {
    for (const pathname of this._localeDirs) {
      try {
        debug(`Loading locales under ${pathname}`)
        await this.loadDirectory(pathname)
      } catch (e) {
        console.error(e)
      }
    }
    if (!this.files.length) {
      throw new Error(`[${PKGNAME}]: No locale files detected. Please check your config.`)
    }
  }

  private async loadDirectory(dirPath: string) {
    const files = await fg('**/*.*', {
      cwd: dirPath,
      onlyFiles: true,
      ignore: ['node_modules/**'],
      deep: undefined,
    })

    for (const relative of files) {
      await this.loadFile(dirPath, relative)
    }
  }

  private async loadFile(dirpath: string, relativePath: string) {
    try {
      const result = this.getFileInfo(dirpath, relativePath)
      if (!result) {
        return
      }
      const { locale, parser, namespace, fullpath: filepath, matcher } = result
      if (!parser) {
        return
      }
      if (!locale) {
        return
      }

      debug(`Loading (${locale}) ${relativePath}`)

      const data = await parser.load(filepath)

      const deepDirpath = path.dirname(filepath)

      this._files[filepath] = {
        filepath,
        deepDirpath,
        dirpath,
        locale,
        value: data,
        namespace,
        matcher,
      }

      return true
    } catch (e) {
      this.unsetFile(relativePath)
      debug(`Failed to load ${e}`)
      console.error(e)
    }
  }

  private unsetFile(filepath: string) {
    delete this._files[filepath]
  }

  private getFileInfo(dirpath: string, relativePath: string) {
    const fullpath = path.resolve(dirpath, relativePath)
    const ext = path.extname(relativePath)

    let match: RegExpExecArray | null = null
    let matcher: string | undefined

    match = this._pathMatcher!.regex.exec(relativePath)
    if (match && match.length > 0) {
      matcher = this._pathMatcher!.matcher
    }

    if (!match || match.length < 1) {
      return
    }

    let namespace = match.groups?.namespace
    if (namespace) {
      namespace = namespace.replaceAll('/', '.')
    }

    const locale = match.groups?.locale

    if (!locale) {
      return
    }

    const parser = this.getMatchedParser(ext)

    return {
      locale,
      parser,
      ext,
      namespace,
      fullpath,
      matcher,
    }
  }

  private getMatchedParser(ext: string) {
    if (!ext.startsWith('.') && ext.includes('.')) {
      ext = path.extname(ext)
    }

    // resolve parser
    return this.getParsers().find((parser) => parser.supports(ext))
  }

  private enabledParserExts() {
    const enabledParsers = this.getParsers().map((item) => item.supportedExts)
    return enabledParsers.filter(Boolean).join('|')
  }

  private getParsers() {
    const _parsers = this.config.parserPlugins?.filter(Boolean)

    const parsers: Parser[] = DefaultEnabledParsers
    if (_parsers?.length) {
      parsers.push(..._parsers.filter(Boolean).map((parser) => new Parser(parser!)))
    }

    return parsers
  }

  get pathMatcher() {
    return this._pathMatcher
  }

  get localeDirs() {
    return this._localeDirs
  }

  async findLocaleDirs() {
    this._files = Object.create(null)
    this._localeModules = Object.create(null)
    this._localeDirs = []

    if (this._localesPaths?.length) {
      try {
        const _locale_dirs = await fg(this._localesPaths, {
          cwd: this._rootPath,
          onlyDirectories: true,
        })

        if (this._localesPaths.includes('.')) {
          _locale_dirs.push('.')
        }

        this._localeDirs = uniq(_locale_dirs.map((p) => path.resolve(this._rootPath, p)))
        debug(`ðŸ“‚ Locale directories:`, this._localeDirs)
      } catch (e) {
        console.error(e)
      }
    }
    if (this._localeDirs.length === 0) {
      logger.error('\nNo locales paths.')
      return false
    }

    return true
  }

  async guessDirStructure(): Promise<DirStructure> {
    const POSITIVE_RATE = 0.6

    const dir = this._localeDirs[0]

    const dirnames = await fg('*', {
      onlyDirectories: true,
      cwd: dir,
      deep: 1,
    })

    const total = dirnames.length
    if (total === 0) return 'file'

    const positives = dirnames.map((d) => tags.check(d))

    const positive = positives.filter((d) => d).length

    // if there are some dirs are named as locale code, guess it's dir mode
    return positive / total >= POSITIVE_RATE ? 'dir' : 'file'
  }

  async resolvePathMatcherByDirStructure() {
    this._dirStructure = await this.guessDirStructure()
    debug(`Directory structure: ${this._dirStructure}`)
    return this.resolvePathMatcher(this._dirStructure)
  }

  resolvePathMatcher(dirStructure?: DirStructure): string {
    if (dirStructure === 'file') {
      return '{locale}.{ext}'
    } else if (this._namespace) {
      return '{locale}/**/{namespace}.{ext}'
    } else {
      return '{locale}/**/*.{ext}'
    }
  }
}
