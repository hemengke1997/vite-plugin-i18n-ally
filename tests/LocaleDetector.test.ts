import path from 'node:path'
import { beforeAll, describe, expect, test } from 'vitest'
import { LocaleDetector } from '../src/plugin/locale-detector/LocaleDetector'
import { initOptions } from '../src/plugin/utils/init-options'

describe('LocaleDetector - Dir mode', () => {
  let localeDetector: LocaleDetector

  beforeAll(async () => {
    const options = initOptions({
      root: path.resolve(__dirname, './fixtures/'),
      localesPaths: [path.resolve(__dirname, './fixtures/locales/')],
      namespace: true,
      dotVscodePath: false,
    })

    localeDetector = new LocaleDetector(options)

    await localeDetector.init()
  })

  test('should localeDetector find locales', () => {
    expect(localeDetector.files.length).toBeGreaterThan(0)
  })

  test('should no loadFileWaitingList', () => {
    expect(localeDetector.loadFileWaitingListLength).toBe(0)
  })

  test('should localeModules has every locale', () => {
    const { modules } = localeDetector.localeModules
    expect(Object.keys(modules).sort()).toMatchInlineSnapshot(`
      [
        "de",
        "en",
        "zh",
        "zh-tw",
      ]
    `)
  })

  test('should virtualModules has every locale', () => {
    const { virtualModules } = localeDetector.localeModules
    expect(Object.keys(virtualModules).sort()).toMatchInlineSnapshot(`
      [
        "virtual:i18n-de",
        "virtual:i18n-en",
        "virtual:i18n-zh",
        "virtual:i18n-zh-tw",
      ]
    `)
  })

  test('should resolvedIds has every locale', () => {
    const { resolvedIds } = localeDetector.localeModules
    const resolvedIdsValues = Array.from(resolvedIds.values()).sort()
    expect(resolvedIdsValues).toMatchInlineSnapshot(`
      [
        "virtual:i18n-de",
        "virtual:i18n-en",
        "virtual:i18n-zh",
        "virtual:i18n-zh-tw",
      ]
    `)
  })

  test('locale modules', () => {
    const { modules } = localeDetector.localeModules
    expect(modules).toMatchSnapshot()
  })

  test('virtualModules', () => {
    const { virtualModules } = localeDetector.localeModules
    expect(virtualModules).toMatchSnapshot()
  })
})
