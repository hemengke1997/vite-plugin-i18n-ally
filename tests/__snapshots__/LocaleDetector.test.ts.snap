// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LocaleDetector - Dir mode > locale modules 1`] = `
{
  "de": {
    "more": {
      "a": {
        "b": {
          "c": "more info",
        },
      },
    },
    "test": {
      "key": "Deutsch",
    },
  },
  "en": {
    "more": {
      "a": {
        "b": {
          "c": "more info default",
        },
      },
      "key": {
        "key-1": "value-1",
      },
    },
    "test": {
      "key": "en",
    },
    "unused": {},
  },
  "zh": {
    "test": {
      "key": "中文",
    },
  },
  "zh-tw": {
    "test": {
      "key": "繁体中文",
    },
  },
}
`;

exports[`LocaleDetector - Dir mode > virtualModules 1`] = `
{
  "virtual:i18n-de": {
    "more": {
      "a": {
        "b": {
          "c": "more info",
        },
      },
    },
    "test": {
      "key": "Deutsch",
    },
  },
  "virtual:i18n-en": {
    "more": {
      "a": {
        "b": {
          "c": "more info default",
        },
      },
      "key": {
        "key-1": "value-1",
      },
    },
    "test": {
      "key": "en",
    },
    "unused": {},
  },
  "virtual:i18n-zh": {
    "test": {
      "key": "中文",
    },
  },
  "virtual:i18n-zh-tw": {
    "test": {
      "key": "繁体中文",
    },
  },
}
`;
